book/db/README.md

```diff
@@ -0,0 +1,332 @@
+```rust ignore
+/// Read write transaction that allows writing to database
+pub trait DbTxMut<'tx>: for<'a> DbTxMutGAT<'a> {
+    /// Put value to database
+    fn put<T: Table>(&self, key: T::Key, value: T::Value) -> Result<(), Error>;
+    /// Delete value from database
+    fn delete<T: Table>(&self, key: T::Key, value: Option<T::Value>) -> Result<bool, Error>;
+    /// Clears database.
+    fn clear<T: Table>(&self) -> Result<(), Error>;
+    /// Cursor mut
+    fn cursor_mut<T: Table>(&self) -> Result<<Self as DbTxMutGAT<'_>>::CursorMut<T>, Error>;
+    /// DupCursor mut.
+    fn cursor_dup_mut<T: DupSort>(
+        &self,
+    ) -> Result<<Self as DbTxMutGAT<'_>>::DupCursorMut<T>, Error>;
+}
+```
+
+Lets take a look at the `DbTx` and `DbTxMut` traits in action. Revisiting the `Transaction` struct as an example, the `Transaction::get_block_hash()` method uses the `DbTx::get()` function to get a block header hash. Remember that the `Transaction` struct's `db` field implements the `Database` trait, which implements the `DatabaseGAT` trait, which defines the `TX` type that implements the `DbTx` trait.
+
+[File: crates/stages/src/db.rs](https://github.com/paradigmxyz/reth/blob/main/crates/stages/src/db.rs#L106)
+```rust ignore
+
+impl<'this, DB> Transaction<'this, DB>
+where
+    DB: Database,
+{
+   //--snip--
+
+    /// Query [tables::CanonicalHeaders] table for block hash by block number
+    pub(crate) fn get_block_hash(&self, number: BlockNumber) -> Result<BlockHash, StageError> {
+        let hash = self
+            .get::<tables::CanonicalHeaders>(number)?
+            .ok_or(DatabaseIntegrityError::CanonicalHash { number })?;
+        Ok(hash)
+    }
+   //--snip--
+}
+```
+
+The `Transaction::get_block_hash()` method takes a `BlockNumber` and returns the block hash at that block number. To do this, the function uses `self.get::<tables::CanonicalHeaders>(number)?.ok_or(DatabaseIntegrityError::CanonicalHash { number })?;`.
+
+Notice that function uses a [turbofish](https://techblog.tonsser.com/posts/what-is-rusts-turbofish) to define which table to use when passing in the `key` to the `DbTx::get()` function. Taking a quick look at the function definition, a generic `T` is defined that implements the `Table` trait mentioned at the beginning of this chapter. 
+
+
+[File: crates/storage/db/src/abstraction/transaction.rs](https://github.com/paradigmxyz/reth/blob/main/crates/storage/db/src/abstraction/transaction.rs#L38)
+```rust ignore
+fn get<T: Table>(&self, key: T::Key) -> Result<Option<T::Value>, Error>;
+```
+
+This design pattern is very powerful and allows Reth to use the methods available to the `DbTx` and `DbTxMut` traits without having to define implementation blocks for each table within the database.  
+
+Lets take a look at a few examples before moving on. In the snippet below, the `DbTx::get()` function is used to get the block header hash at the specified `block_num` from the `CanonicalHeaders` table.
```

Comment:
not sure if this example is necessary, it's very similar to the one above that references `Transaction::get_block_hash`